java/ql/src/semmle/code/java/security/XPath.qll

```diff
@@ -0,0 +1,65 @@
+/** Provides classes to reason about XPath vulnerabilities. */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+
+/**
+ * An abstract type representing a call to interpret XPath expressions.
+ */
+class XPathSink extends MethodAccess {
+  /**
+   * Gets the argument representing the XPath expressions to be evaluated.
+   */
+  abstract Expr getSink();
+}
+
+/** The class `javax.xml.xpath.XPath` */
+class XPath extends RefType {
+  XPath() { this.hasQualifiedName("javax.xml.xpath", "XPath") }
+}
+
+/** A call to `XPath.evaluate` or `XPath.compile` */
+class XPathEvaluateOrCompile extends XPathSink {
+  XPathEvaluateOrCompile() {
+    exists(Method m | this.getMethod() = m and m.getDeclaringType() instanceof XPath |
```

Comment:
Would it make sense to check for `getDeclaringType().getASourceSupertype*() instanceof XPath` here in case there are public classes implementing the method?

Reply:
I would say that most of the time applications will use the `XPath` interface and not a specific implementation, but it seems like a good addition to catch potential edge cases. Thanks! Added in ab543de